cli/cmd/logs.go

```diff
@@ -0,0 +1,224 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+//This code replicates most of the functionality in https://github.com/wercker/stern/blob/master/cmd/cli.go
+type logCmdConfig struct {
+	clientset *kubernetes.Clientset
+	*stern.Config
+}
+
+type logsOptions struct {
+	container             string
+	controlPlaneComponent string
+	sinceSeconds          time.Duration
+	tail                  int64
+	timestamps            bool
+}
+
+func newLogsOptions() *logsOptions {
+	return &logsOptions{
+		container:             "",
+		controlPlaneComponent: "",
+		sinceSeconds:          48 * time.Hour,
+		tail:                  -1,
+		timestamps:            false,
+	}
+}
+
+func (o *logsOptions) toSternConfig(controlPlaneComponents, availableContainers []string) (*stern.Config, error) {
+	config := &stern.Config{}
+
+	if o.controlPlaneComponent == "" {
+		config.LabelSelector = labels.Everything()
+	} else {
+		var podExists string
+		for _, p := range controlPlaneComponents {
+			if p == o.controlPlaneComponent {
+				podExists = p
+				break
+			}
+		}
+
+		if podExists == "" {
+			return nil, fmt.Errorf("control plane component [%s] does not exist. Must be one of %v", o.controlPlaneComponent, controlPlaneComponents)
+		}
+		selector, err := labels.Parse(fmt.Sprintf("linkerd.io/control-plane-component=%s", o.controlPlaneComponent))
+		if err != nil {
+			return nil, err
+		}
+		config.LabelSelector = selector
+	}
+
+	if o.container != "" {
+		var matchingContainer string
+		for _, c := range availableContainers {
+			if o.container == c {
+				matchingContainer = c
+				break
+			}
+		}
+		if matchingContainer == "" {
+			return nil, fmt.Errorf("container [%s] does not exist in control plane [%s]", o.container, controlPlaneNamespace)
+		}
+	}
+
+	containerFilterRgx, err := regexp.Compile(o.container)
+	if err != nil {
+		return nil, err
+	}
+	config.ContainerQuery = containerFilterRgx
+
+	if o.tail != -1 {
+		config.TailLines = &o.tail
+	}
+
+	// Do not use regex to filter pods. Instead, we provide the list of all control plane components and use
+	// the label selector to filter logs.
+	podFilterRgx, err := regexp.Compile("")
+	if err != nil {
+		return nil, err
+	}
+	config.PodQuery = podFilterRgx
+	config.Since = o.sinceSeconds
+	config.Timestamps = o.timestamps
+	config.Namespace = controlPlaneNamespace
+
+	return config, nil
+}
+
+func getControlPlaneComponentsAndContainers(pods *v1.PodList) ([]string, []string) {
+	var controlPlaneComponents, containers []string
+	for _, pod := range pods.Items {
+		controlPlaneComponents = append(controlPlaneComponents, pod.Labels["linkerd.io/control-plane-component"])
+		for _, container := range pod.Spec.Containers {
+			containers = append(containers, container.Name)
+		}
+	}
+	return controlPlaneComponents, containers
+}
+
+func newLogCmdConfig(options *logsOptions, kubeconfigPath, kubeContext string) (*logCmdConfig, error) {
+	// Check that we can call the Kubernetes API
+	_ := cliPublicAPIClient()
```

Comment:
I'm actually on the fence about needing this validation. It seems like the `linkerd logs` command might be most helpful when the control plane is not functioning as expected. And in that case we may not successfully be able to connect to the API, and the command will exit with an error here. But I could see the value in fetching container logs, even if the public API is inaccessible. So maybe just omit this?

Reply:
Very good point. TIL. I thought calling`cliPublicAPIClient()` without any assignment variable would lead other readers to believe that the method was side-effect-y but it looks like it goes against the `go vet`'s suggestion. 

Also good point on the check it may not be necessary. 👍